### PR TITLE
Kitography/vault 5474

### PR DESCRIPTION
**Pull request by kitography**
There's a bug in PEM/DER encoding of private keys here, but creating the long overdue PR.

### CODE REVIEW

**Review by stevendpclark — COMMENTED** [review]
Overall great stuff, a few nits and comments to possibly address.

**Comment by kitography** [comment]
Closing this - i really need to rebase, and made a mess of this PR not realizing I couldn't do that here.  (Sorry for the spam ecosystem + taoism4504) https://github.com/hashicorp/vault/pull/15150

File: builtin/logical/pki/path_fetch_keys.go [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+			},
+		},
+
+		HelpSynopsis:    pathListKeysHelpSyn,
+		HelpDescription: pathListKeysHelpDesc,
+	}
+}
+
+const (
+	pathListKeysHelpSyn  = ``
```
[the remark this answers is not in the document]
Comment (kitography) [reply builtin/logical/pki/path_fetch_keys.go:29]: I think I've fixed all of these (in the rebased PR), but let me know if I've missed any there - this one is definitely fixed. :)

File: sdk/framework/backend_test.go [hunk 2]
```diff
@@ -2,15 +2,16 @@ package framework
...
-	"github.com/hashicorp/go-secure-stdlib/strutil"
-	"github.com/stretchr/testify/require"
 	"net/http"
 	"reflect"
 	"strings"
 	"sync/atomic"
 	"testing"
 	"time"
 
+	"github.com/hashicorp/go-secure-stdlib/strutil"
```
[the remark this answers is not in the document]
Comment (kitography) [reply sdk/framework/backend_test.go:12]: It was, but it's changed every time I run "make fmt" in order to fix the test-colors.

File: builtin/logical/pki/path_manage_keys.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+	keyBits := data.Get("key_bits").(int)
+	switch {
+	case strings.HasSuffix(req.Path, "/internal"):
+		// Internal key generation, stored in storage
+		keyBundle, err := certutil.GetKeyBundleFromKeyGenerator(keyType, keyBits, nil)
+		if err != nil {
+			return nil, err
+		}
+		importKey(ctx, req.Storage, string(keyBundle.PrivateKeyBytes), keyName)
+	case strings.HasSuffix(req.Path, "/exported"):
```
[the remark this answers is not in the document]
Comment (kitography) [reply builtin/logical/pki/path_manage_keys.go:64]: yeah, this was a mistake, I've fixed that in the rebased PR - thanks for explaining.

File: builtin/logical/pki/path_manage_keys.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+		keyRef := data.Get("key_ref").(string)
+		keyBundle, err := certutil.GetKeyBundleFromKeyGenerator(keyType, keyBits, existingGeneratePrivateKey(ctx, req.Storage, keyRef))
+		if err != nil {
+			return nil, err
+		}
+		importKey(ctx, req.Storage, string(keyBundle.PrivateKeyBytes), keyName)
+	default:
```
[the remark this answers is not in the document]
Comment (kitography) [reply builtin/logical/pki/path_manage_keys.go:71]: yep, thank you - we'll need to do this on ent, but I've added a "ent only" error on the rebased PR.

File: sdk/helper/certutil/types.go [hunk 5]
```diff
@@ -138,6 +138,13 @@ type ParsedCSRBundle struct {
 	CSR             *x509.CertificateRequest
 }
 
+type KeyBundle struct {
+	PrivateKeyType   PrivateKeyType
+	PrivateKeyBytes  []byte
+	PrivateKeyString string
```
[the remark this answers is not in the document]
Comment (kitography) [reply sdk/helper/certutil/types.go:144]: We can - unfortunately I need a method to fetch the PEM string - so I've removed this, and added that method to the rebased PR.

File: builtin/logical/pki/path_fetch_keys.go [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+	pattern := "key/" + framework.GenericNameRegex("ref")
+	return buildPathKey(b, pattern)
+}
+
+func buildPathKey(b *backend, pattern string) *framework.Path {
+	return &framework.Path{
+		Pattern: pattern,
+
+		Fields: map[string]*framework.FieldSchema{
+			"ref": {
```
[the remark this answers is not in the document]
Comment (kitography) [reply builtin/logical/pki/path_fetch_keys.go:67]: I've fixed this (just now) in the new PR - I think it's a good idea, and it makes using some of your functions which assume position in certain places of data much easier to work with.

File: builtin/logical/pki/path_fetch_keys.go [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+
+	keyId, err := resolveKeyReference(ctx, req.Storage, keyRef)
+	if err != nil {
+		return nil, err
+	}
+	if keyId == "" {
+		return logical.ErrorResponse("unable to resolve key id for reference" + keyRef), nil
+	}
+
+	wasDefault, err := deleteKey(ctx, req.Storage, keyId)
```
[the remark this answers is not in the document]
Comment (kitography) [reply builtin/logical/pki/path_fetch_keys.go:198]: Oh, oops.  Okay, fixed that on new PR.

**Pull request closed without merging**